File: komodo/pr.py
"""GitHub pull request actions through the gh CLI, each scoped to one branch's own PR."""

from __future__ import annotations

import json
import re
import subprocess
from typing import Any, Dict, List, Optional


class GhError(RuntimeError):
    """gh exited non-zero or returned something unparseable."""


def gh(*args: str, cwd: str, timeout: int = 120, check: bool = True) -> str:
    """Runs gh and returns stdout."""
    result = subprocess.run(["gh", *args], cwd=cwd, capture_output=True, text=True, timeout=timeout)
    if check and result.returncode != 0:
        raise GhError("gh %s failed: %s" % (" ".join(args[:3]), result.stderr.strip()[:600]))
    return result.stdout


def gh_json(*args: str, cwd: str) -> Any:
    """Runs gh and parses its JSON output."""
    out = gh(*args, cwd=cwd)
    try:
        return json.loads(out) if out.strip() else None
    except ValueError:
        raise GhError("gh returned non-JSON for %s" % " ".join(args[:3]))


def available(cwd: str) -> bool:
    """Whether gh is installed and authenticated."""
    try:
        return subprocess.run(["gh", "auth", "status"], cwd=cwd, capture_output=True, timeout=30).returncode == 0
    except (OSError, subprocess.SubprocessError):
        return False


def repo_slug(cwd: str) -> str:
    """owner/name for the current repo."""
    data = gh_json("repo", "view", "--json", "nameWithOwner", cwd=cwd)
    return str(data["nameWithOwner"])


def view(cwd: str, branch: Optional[str] = None) -> Optional[Dict[str, Any]]:
    """The PR for a branch (default: current), or None when there is none."""
    args = ["pr", "view", "--json", "number,url,state,title,labels,headRefName,baseRefName,mergedAt"]
    if branch:
        args.append(branch)
    try:
        return gh_json(*args, cwd=cwd)
    except GhError:
        return None


def create(cwd: str, title: str, body: str, base: str, labels: List[str], draft: bool = False) -> str:
    """Opens a PR for the current branch and returns its URL."""
    args = ["pr", "create", "--title", title, "--body", body, "--base", base]
    for label in labels:
        args += ["--label", label]
    if draft:
        args.append("--draft")
    return gh(*args, cwd=cwd, timeout=180).strip()


def edit(cwd: str, number: int, body: Optional[str] = None, title: Optional[str] = None, add_labels: List[str] = ()) -> None:
    """Updates the body, title, or labels of a PR the branch owns."""
    args = ["pr", "edit", str(number)]
    if body is not None:
        args += ["--body", body]
    if title is not None:
        args += ["--title", title]
    for label in add_labels:
        args += ["--add-label", label]
    gh(*args, cwd=cwd)


def existing_labels(cwd: str) -> List[str]:
    """Label names the repo defines."""
    data = gh_json("label", "list", "--json", "name", "--limit", "200", cwd=cwd) or []
    return [str(item["name"]) for item in data]


# A conventional-commit subject: the type, an optional scope, then the colon.
CONVENTIONAL = re.compile(r"^([a-z]+)(\([^)]*\))?!?:")


def kind_of(title: str, branch: str) -> str:
    """The commit type a PR carries, read from its title, then from its branch prefix."""
    match = CONVENTIONAL.match(title.strip())
    if match:
        return match.group(1)
    head = branch.split("/", 1)[0].strip().lower()
    return head if head and head != branch else ""


def pick_labels(kind: str, mapping: Dict[str, str], defined: List[str]) -> List[str]:
    """Category and authorship labels that exist in the repo, never inventing one."""
    chosen = []
    category = mapping.get(kind)
    if category and category in defined:
        chosen.append(category)
    agent = mapping.get("agent")
    if agent and agent in defined:
        chosen.append(agent)
    return chosen


def comment(cwd: str, number: int, body: str) -> None:
    """Posts a top-level PR comment."""
    gh("pr", "comment", str(number), "--body", body, cwd=cwd)


def reply(cwd: str, number: int, comment_id: int, body: str) -> None:
    """Replies inside a review thread; gh pr comment has no thread mode, so this is the one raw API write."""
    slug = repo_slug(cwd)
    gh("api", "-X", "POST", "repos/%s/pulls/%d/comments/%d/replies" % (slug, number, comment_id), "-f", "body=%s" % body, cwd=cwd)


def review_threads(cwd: str, number: int) -> List[Dict[str, Any]]:
    """Unresolved review threads with their comments, via GraphQL."""
    slug = repo_slug(cwd)
    owner, name = slug.split("/", 1)
    query = (
        "query($owner:String!,$name:String!,$number:Int!){repository(owner:$owner,name:$name){pullRequest(number:$number){"
        "reviewThreads(first:50){nodes{id isResolved path line comments(first:20){nodes{databaseId author{login} body createdAt}}}}}}}"
    )
    data = gh_json("api", "graphql", "-f", "query=%s" % query, "-F", "owner=%s" % owner, "-F", "name=%s" % name, "-F", "number=%d" % number, cwd=cwd)
    nodes = (((data or {}).get("data") or {}).get("repository") or {}).get("pullRequest", {}).get("reviewThreads", {}).get("nodes", [])
    threads = []
    for node in nodes:
        if node.get("isResolved"):
            continue
        comments = [
            {"id": item.get("databaseId"), "author": (item.get("author") or {}).get("login", ""), "body": item.get("body", ""), "at": item.get("createdAt", "")}
            for item in node.get("comments", {}).get("nodes", [])
        ]
        threads.append({"id": node.get("id"), "path": node.get("path") or "", "line": node.get("line") or 0, "comments": comments})
    return threads


def merged_prs_for(cwd: str, branches: List[str]) -> List[str]:
    """Which of the given branches have a merged PR."""
    merged = []
    for branch in branches:
        info = view(cwd, branch)
        if info and info.get("state") == "MERGED":
            merged.append(branch)
    return merged

File: internal/pr/pr.go
// Package pr wraps the gh commands the line uses to open and answer pull requests.
package pr

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os/exec"
	"strings"
)

// Thread is one unresolved review comment on a pull request.
type Thread struct {
	ID     string `json:"id"`
	Path   string `json:"path"`
	Line   int    `json:"line"`
	Author string `json:"author"`
	Body   string `json:"body"`
}

// Pull is the fields the line reads back about a pull request.
type Pull struct {
	Number int    `json:"number"`
	URL    string `json:"url"`
	State  string `json:"state"`
	Title  string `json:"title"`
	Draft  bool   `json:"isDraft"`
}

// Runner runs one gh invocation, so a test can stand in for the real CLI.
type Runner func(dir string, args ...string) (string, error)

// Run is the default runner: the gh binary on PATH.
func Run(dir string, args ...string) (string, error) {
	cmd := exec.Command("gh", args...)
	cmd.Dir = dir
	var stdout, stderr bytes.Buffer
	cmd.Stdout, cmd.Stderr = &stdout, &stderr
	if err := cmd.Run(); err != nil {
		return "", fmt.Errorf("gh %s: %v: %s", strings.Join(args, " "), err, strings.TrimSpace(stderr.String()))
	}
	return strings.TrimSpace(stdout.String()), nil
}

// Client talks to one repository through gh.
type Client struct {
	Dir string
	Run Runner
}

// New returns a client rooted at dir using the gh binary.
func New(dir string) *Client { return &Client{Dir: dir, Run: Run} }

// run invokes the client's runner.
func (c *Client) run(args ...string) (string, error) {
	runner := c.Run
	if runner == nil {
		runner = Run
	}
	return runner(c.Dir, args...)
}

// Create opens a pull request and returns its URL.
func (c *Client) Create(base, head, title, body string, draft bool) (string, error) {
	args := []string{"pr", "create", "--base", base, "--head", head, "--title", title, "--body", body}
	if draft {
		args = append(args, "--draft")
	}
	return c.run(args...)
}

// View returns what gh knows about one pull request, or the current branch's.
func (c *Client) View(number string) (Pull, error) {
	var pull Pull
	args := []string{"pr", "view", "--json", "number,url,state,title,isDraft"}
	if number != "" {
		args = append(args[:2], append([]string{number}, args[2:]...)...)
	}
	out, err := c.run(args...)
	if err != nil {
		return pull, err
	}
	return pull, json.Unmarshal([]byte(out), &pull)
}

// Edit changes a pull request's body, title, or draft state.
func (c *Client) Edit(number string, args ...string) error {
	_, err := c.run(append([]string{"pr", "edit", number}, args...)...)
	return err
}

// Label adds labels to a pull request, ignoring ones the repo does not define.
func (c *Client) Label(number string, labels []string) error {
	if len(labels) == 0 {
		return nil
	}
	args := []string{"pr", "edit", number}
	for _, label := range labels {
		args = append(args, "--add-label", label)
	}
	_, err := c.run(args...)
	return err
}

// Labels lists the labels the repository already defines.
func (c *Client) Labels() ([]string, error) {
	out, err := c.run("label", "list", "--json", "name")
	if err != nil {
		return nil, err
	}
	var rows []struct {
		Name string `json:"name"`
	}
	if err := json.Unmarshal([]byte(out), &rows); err != nil {
		return nil, err
	}
	names := make([]string, 0, len(rows))
	for _, row := range rows {
		names = append(names, row.Name)
	}
	return names, nil
}

// Comment posts one comment on a pull request.
func (c *Client) Comment(number, body string) error {
	_, err := c.run("pr", "comment", number, "--body", body)
	return err
}

// threadsQuery fetches a pull request's inline review threads by URL.
const threadsQuery = `query($url: URI!) {
  resource(url: $url) {
    ... on PullRequest {
      reviewThreads(first: 100) {
        nodes {
          id
          isResolved
          path
          line
          comments(first: 10) {
            nodes {
              body
              author { login }
            }
          }
        }
      }
    }
  }
}`

// replyMutation adds a reply comment inside one review thread.
const replyMutation = `mutation($id: ID!, $body: String!) {
  addPullRequestReviewThreadReply(input: {pullRequestReviewThreadId: $id, body: $body}) {
    comment { id }
  }
}`

// Threads lists the unresolved review threads on a pull request.
func (c *Client) Threads(number string) ([]Thread, error) {
	pull, err := c.View(number)
	if err != nil {
		return nil, err
	}
	out, err := c.run("api", "graphql", "-f", "query="+threadsQuery, "-f", "url="+pull.URL)
	if err != nil {
		return nil, err
	}
	var payload struct {
		Data struct {
			Resource struct {
				ReviewThreads struct {
					Nodes []struct {
						ID         string `json:"id"`
						IsResolved bool   `json:"isResolved"`
						Path       string `json:"path"`
						Line       int    `json:"line"`
						Comments   struct {
							Nodes []struct {
								Body   string `json:"body"`
								Author struct {
									Login string `json:"login"`
								} `json:"author"`
							} `json:"nodes"`
						} `json:"comments"`
					} `json:"nodes"`
				} `json:"reviewThreads"`
			} `json:"resource"`
		} `json:"data"`
	}
	if err := json.Unmarshal([]byte(out), &payload); err != nil {
		return nil, err
	}
	var threads []Thread
	for _, node := range payload.Data.Resource.ReviewThreads.Nodes {
		if node.IsResolved || len(node.Comments.Nodes) == 0 {
			continue
		}
		opening := node.Comments.Nodes[0]
		threads = append(threads, Thread{
			ID: node.ID, Path: node.Path, Line: node.Line,
			Author: opening.Author.Login, Body: opening.Body,
		})
	}
	return threads, nil
}

// Reply answers one review thread by posting inside it, not as a new top-level comment.
func (c *Client) Reply(threadID, body string) error {
	_, err := c.run("api", "graphql", "-f", "query="+replyMutation, "-f", "id="+threadID, "-f", "body="+body)
	return err
}

// resolveMutation marks one review thread resolved.
const resolveMutation = `mutation($id: ID!) {
  resolveReviewThread(input: {threadId: $id}) {
    thread { id }
  }
}`

// Resolve marks one review thread resolved, so it drops out of the next Threads call.
func (c *Client) Resolve(threadID string) error {
	_, err := c.run("api", "graphql", "-f", "query="+resolveMutation, "-f", "id="+threadID)
	return err
}

// KeepKnown returns only the labels the repository already defines.
func KeepKnown(wanted, known []string) []string {
	var out []string
	for _, label := range wanted {
		for _, candidate := range known {
			if label == candidate {
				out = append(out, label)
				break
			}
		}
	}
	return out
}
